File: 357. Count Numbers with Unique Digits.cpp
class Solution {
public:
    // math & dp: O(1)time & space, 13.25%, ok
    int countNumbersWithUniqueDigits(int n) {
        if(!n) return 1;  // corner case
        int ret = 10;     // when length = 1
        int multi = 9;
        int product = 9;
        while(n > 1 && multi > 0){  // when length > 10, no longer changes
            product *= multi--;
            ret += product;
            n--;
        }
        return ret;
    }

};

File: 357. Count Numbers with Unique Digits.py
class Solution(object):
    
    # backtracking, dfs: O() time
    def countNumbersWithUniqueDigits(self, n):
        """
        :type n: int
        :rtype: int
        """
        def dfs(nums, index, length):
            if index == length:
                self.ret += 1
                return 
            for i in xrange(index, 10):
                nums[i], nums[index] = nums[index], nums[i]
                dfs(nums, index + 1, length)
                nums[i], nums[index] = nums[index], nums[i]
        if n == 0:
            return 1
        self.ret = 10
        for length in xrange(2, min(10, n) + 1):  # when n > 10, no valid number 
            dfs(range(10), 0, length)
            self.ret -= reduce(operator.mul, [x for x in xrange(9, 10 - length, -1)])  # leading zero
        return self.ret
                
            
                
            
        